### ISSUE TITLE
rstudio ide hangs irregularly (rsession process 100%)

### ISSUE THREAD

**Issue by lbergmans** [rstudio/rstudio#6645]
For a while, I am experiencing frequent, but unpredictable and irregular hangups of my rstudio environment.
These require me to force-quit (I'm on a Mac, 10.14.6 Mojave) Rstudio and restart. The UI is still operational, but no actual actions are executed). I typically notice it when trying to execute some R code, or e.g. trying to save the file I am editing (the source code editor itself keeps working fine). Whenever I run or knit an entire notebook, or source a file, my code runs and finishes just fine, so I'm pretty sure the problem is not in my own code (..)

Due to multiple cores, I did not realize it, but it appears that rsession is eating 99-100% CPU, probably in a loop?
Other errors that may appear at these times:

error: node stack overflow (in the Console)
error pop-ups such as "Error retrieving help--R code execution error"
".... --unable to establish connection with R session"

I have tried a wide range of things:
* make sure DB connections are closed, later not using DB access at all, but problem remained
* removed tidylog, 
* updated all packages I am using
* reset Rstudio settings: https://support.rstudio.com/hc/en-us/articles/200534577-Resetting-RStudio-Desktop-s-State
  - mv ~/.rstudio-desktop ~/backup-rstudio-desktop
  - defaults delete com.rstudio.desktop
  - mv .Rproj.user .Rproj.user-BAK
  - rm .Rhistory
  - mv .RData .Rdata.BAK
* created a new project
* do no longer save session data
* turned off background saving of source files
* reboot machine
* reinstall R and rstudio
* installed Rstudio v1.3.919-2 Preview
* turned on  software rendering

It is getting really annoying to be interrupted several times a day! (need to force-quite rstudio and restart, and restore my last changes to the source I am working on at the time) Any help/suggestion is appreciated.

below is my session info (from an early moment when this problem occurred), and I attached an rstudio-sample file.

cheers,
Lodewijk Bergmans

-------------------
sessionInfo()
R version 3.6.3 (2020-02-29)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats graphics grDevices utils datasets methods base

other attached packages:
[1] SIG.R.Theme_2020.03.03 interp_1.0-33 lattice_0.20-38 lubridate_1.7.4 knitr_1.28
[6] glue_1.3.2 treemapify_2.5.3 scales_1.1.0 gghighlight_0.2.0 RMariaDB_1.0.8
[11] forcats_0.5.0 stringr_1.4.0 dplyr_0.8.5 purrr_0.3.3 readr_1.3.1
[16] tidyr_1.0.2 tibble_2.1.3 ggplot2_3.3.0 tidyverse_1.3.0

loaded via a namespace (and not attached):
[1] tidyselect_1.0.0 xfun_0.12 haven_2.2.0 colorspace_1.4-1 vctrs_0.2.4 generics_0.0.2
[7] yaml_2.2.1 rlang_0.4.5 pillar_1.4.3 withr_2.1.2 DBI_1.1.0 RcppZiggurat_0.1.5
[13] bit64_0.9-7 dbplyr_1.4.2 modelr_0.1.6 readxl_1.3.1 lifecycle_0.2.0 ggfittext_0.8.1
[19] munsell_0.5.0 gtable_0.3.0 cellranger_1.1.0 rvest_0.3.5 labeling_0.3 parallel_3.6.3
[25] fansi_0.4.1 Rfast_1.9.9 broom_0.5.5 Rcpp_1.0.4 backports_1.1.5 jsonlite_1.6.1
[31] farver_2.0.3 deldir_0.1-25 fs_1.3.2 bit_1.1-15.2 digest_0.6.25 hms_0.5.3
[37] stringi_1.4.6 ggrepel_0.8.2 grid_3.6.3 cli_2.0.2 tools_3.6.3 magrittr_1.5
[43] crayon_1.3.4 pkgconfig_2.0.3 ellipsis_0.3.0 xml2_1.2.5 reprex_0.3.0 assertthat_0.2.1
[49] httr_1.4.1 rstudioapi_0.11 R6_2.4.1 nlme_3.1-144 compiler_3.6.3

**Comment by ronblum** [comment]
@lbergmans Thank you for raising the issue. I see that you've tried resolving this by installing the 1.3 Middlemist Red preview. Which version were you using beforehand?

In any case, I can't reproduce this in 1.2 or 1.3, so I'll mark this for triage to see what other information could be useful.

**Comment by jmcphers** [comment]
Could you grab a sample of the rsession process while it is consuming all that CPU? Here's how:

https://gist.github.com/loderunner/36724cc9ee8db66db305

**Comment by lbergmans** [comment]
thanks for picking this up, I thought I had attached a sample, but apparently did not.. Here's another attempt..
(if I encounter the issue again later this week, I can do another sampling if needed)

Before using the preview version, I was on v1.2.5033, the latest at the time  (I see we are now on 1.2.5042--any chance that would make a difference?)

BTW: I first brought up the issue in this thread: https://community.rstudio.com/t/intermittent-rstudio-hangup-rsession-at-100-cpu/56931, where one more person chimed in with a similar/the same problem.
[rstudio-sample.txt](https://github.com/rstudio/rstudio/files/4471575/rstudio-sample.txt)

**Comment by jmcphers** [comment]
No, 5033 and 5042 are very similar and don't have any changes that would affect this.

That sample appears to be from the RStudio binary, but you mentioned earlier that it was rsession that was hanging, so if that's indeed true there isn't much we can glean from it. :-/ Can you get a sample from rsession?

**Comment by lbergmans** [comment]
ok, I'll start the project, and then wait until the problem re-appears and run a sample.

BTW: I was first a bit confused by your link above, since I am often using the alias 'loderunner' myself.. (but not on github)

**Comment by jmcphers** [comment]
Is it because of [Lode Runner](https://en.wikipedia.org/wiki/Lode_Runner)? That game was legendary!

**Comment by lbergmans** [comment]
euh, yes, but to be honest, I never played it.. 
But I am a bit of a runner, and my first name is 'Lodewijk'.

**Comment by lbergmans** [comment]
well, I actually had to work on the same project today; and i had no issue at all, until 17:30. So here's a sample, obtained with `sample <rsession PID> 30 -file rsession.sample.txt`; 
I hope it is useful: 
[rsession.sample.txt](https://github.com/rstudio/rstudio/files/4476300/rsession.sample.txt)

**Comment by kevinushey** [comment]
It looks like there's some heavily recursive R calls being made, eventually routing to `printf` and similar string handling / printing routines. Unfortunately, we don't have the R call stack so it's tough to know exactly what's going on.

FWIW, from the trace:

```
Total number in stack (recursive counted multiple, when >=5):
        535       Rf_eval  (in libR.dylib) + 445  [0x106b22a7d]  eval.c:620
        521       bcEval  (in libR.dylib) + 27889  [0x106b29ea1]  eval.c:6733
        520       R_execClosure  (in libR.dylib) + 2153  [0x106b41339]  eval.c:0
        520       Rf_applyClosure  (in libR.dylib) + 346  [0x106b4023a]  eval.c:1706
        24       Rprintf  (in libR.dylib) + 144  [0x106bafb50]  printutils.c:847
        18       _platform_strcmp  (in libsystem_platform.dylib) + 0  [0x7fff662b0360]
        17       Rvprintf  (in libR.dylib) + 145  [0x106bafc01]  printutils.c:930
        15       _platform_strlen  (in libsystem_platform.dylib) + 0  [0x7fff662af6e0]
        15       forcePromise  (in libR.dylib) + 179  [0x106b3f9f3]  eval.c:516
        14       R_CheckUserInterrupt  (in libR.dylib) + 77  [0x106b1b1cd]  errors.c:131
        14       Rf_eval  (in libR.dylib) + 580  [0x106b22b04]  eval.c:656
        14       Rvprintf  (in libR.dylib) + 71  [0x106bafbb7]  printutils.c:922
        14       bcEval  (in libR.dylib) + 28934  [0x106b2a2b6]  eval.c:6765
        13       ???  (in rsession)  load address 0x105e02000 + 0x8daf12  [0x1066dcf12]
        13       R_ProcessEvents  (in libR.dylib) + 55  [0x106c66437]  sys-unix.c:883
        13       do_cat  (in libR.dylib) + 2005  [0x106ab0125]  builtin.c:679
        13       do_cat  (in libR.dylib) + 2264  [0x106ab0228]  builtin.c:519
        12       bcEval  (in libR.dylib) + 21753  [0x106b286a9]  eval.c:6681
        12       text_vfprintf  (in libR.dylib) + 351  [0x106adbc0f]  connections.c:3161
        10       ???  (in rsession)  load address 0x105e02000 + 0x104b39  [0x105f06b39]
        10       ???  (in rsession)  load address 0x105e02000 + 0x1930f  [0x105e1b30f]
        10       vsnprintf  (in libsystem_c.dylib) + 68  [0x7fff66161baa]
        9       ???  (in rsession)  load address 0x105e02000 + 0x1941e  [0x105e1b41e]
        8       Rf_allocVector3  (in libR.dylib) + 1648  [0x106b7b0b0]  memory.c:2767
        8       _pthread_body  (in libsystem_pthread.dylib) + 126  [0x7fff662bb2eb]
        8       _pthread_start  (in libsystem_pthread.dylib) + 66  [0x7fff662be249]
        8       _vsnprintf  (in libsystem_c.dylib) + 410  [0x7fff66161b03]
        8       bcEval  (in libR.dylib) + 26793  [0x106b29a59]  eval.c:6728
        8       malloc  (in libsystem_malloc.dylib) + 24  [0x7fff66271c15]
        8       malloc_zone_malloc  (in libsystem_malloc.dylib) + 103  [0x7fff66271c99]
        8       thread_start  (in libsystem_pthread.dylib) + 13  [0x7fff662ba40d]
        7       ???  (in rsession)  load address 0x105e02000 + 0x975198  [0x106777198]
        7       Rstrlen  (in libR.dylib) + 97  [0x106bae971]  printutils.c:507
        7       bcEval  (in libR.dylib) + 29417  [0x106b2a499]  eval.c:6785
        7       text_vfprintf  (in libR.dylib) + 787  [0x106adbdc3]  connections.c:3191
        6       Rf_xlengthgets  (in libR.dylib) + 189  [0x106ab0e9d]  builtin.c:846
        6       Rstrwid  (in libR.dylib) + 491  [0x106bae54b]  printutils.c:413
        6       __v2printf  (in libsystem_c.dylib) + 475  [0x7fff6617c1c3]
        6       _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 0  [0x7fff662afcc0]
        6       do_internal  (in libR.dylib) + 339  [0x106b84733]  names.c:1376
        6       gmtsub  (in libsystem_c.dylib) + 31  [0x7fff6616487d]
        6       text_vfprintf  (in libR.dylib) + 304  [0x106adbbe0]  connections.c:3158
        6       text_vfprintf  (in libR.dylib) + 895  [0x106adbe2f]  connections.c:3192
        5       Rf_install  (in libR.dylib) + 127  [0x106b8417f]  names.c:1231
        5       Rf_mkCharLenCE  (in libR.dylib) + 727  [0x106b19af7]  envir.c:3964
        5       bcEval  (in libR.dylib) + 15988  [0x106b27024]  eval.c:6517
        5       do_cat  (in libR.dylib) + 2569  [0x106ab0359]  builtin.c:691
        5       notify_check_tz  (in libsystem_c.dylib) + 44  [0x7fff66163340]
        5       szone_malloc_should_clear  (in libsystem_malloc.dylib) + 120  [0x7fff66273245]
```

**Comment by lbergmans** [comment]
Would there be any trace in here that points to any user (my) R code?
It is hard to imagine, because it would mean that after my code has finished, it leaves some active threads behind (I am not explicitly creating any threads in my code, but perhaps indirectly..)

**Comment by lbergmans** [comment]
just curious; where do we go from here? is any follow-up planned?

**Comment by kevinushey** [comment]
Unfortunately, without a reproducible example there's not much more we can do.

**Comment by lbergmans** [comment]
ok, so the trace is about the right process, but has not enough info?
I can understand that you want/need to see the actual problem to be able to solve it.
wish I could just produce/predict when it occurs..

**Comment by kevinushey** [comment]
Right, that's correct. Definitely understand that it can be challenging to get a reproducible example, especially since some issues (perhaps this one) occur non-deterministically, but usually there is some sequence of actions that reproduces (or increases the chances of reproducing) the issue.

**Comment by lbergmans** [comment]
that makes sense, and I have the feeling that this problem occurs mostly in this specific project I have been working on, but I have tried to rule out all 'special' aspects such as DB access, but it has not given me any more insights.. If I find out about specific circumstances that seem to play a role, I'll report those. Thanks for trying and all the swift responses!

**Comment by mpjdem** [comment]
I have the same problem, using Ubuntu 20.04 and RStudio 1.3.947. I was experiencing it before on 19.10 and with older RStudio versions as well, for the past number of weeks.

The rsession process will go to 100% for perhaps 10-15 minutes, when eventually a pop-up box appears 'Error retrieving help', together with 'Error: C stack usage  7975060 is too close to the limit' in the console.

It does seem to occur using a specific project, where I use devtools::load_all() to load the functions in R/. The problem is not obviously reactive to me doing anything, and occurs even when no objects are present in the global environment. I will usually notice it because my files won't save or there is no code completion.

**Comment by mpjdem** [comment]
@kevinushey @jmcphers I am now able to reliably reproduce this. The culprit in my case seems to be the package 'Rfast'. When this package's namespace is loaded, it seems to interfere with autocompletion of dplyr functions in RStudio. In a terminal R session there is no problem.

Below a Dockerfile where I get the error every time by doing the following steps in the RStudio Server console:
    - library(Rfast)
    - dplyr::filter( <TAB> 

Sometimes it takes more than one attempt to get the bug, but usually not. Hope this helps!

`FROM rocker/r-ver:latest

RUN apt-get update && apt-get install -y sudo wget psmisc libapparmor1 libclang-dev libgsl-dev

RUN wget -q https://s3.amazonaws.com/rstudio-ide-build/server/bionic/amd64/rstudio-server-1.3.957-amd64.deb && dpkg -i rstudio-server-1.3.957-amd64.deb

RUN Rscript -e "install.packages('dplyr')"
RUN Rscript -e "install.packages('Rfast')"

RUN useradd rstudio \
  && echo "rstudio:rstudio" | chpasswd \
  && mkdir /home/rstudio \
  && chown rstudio:rstudio /home/rstudio \
  && addgroup rstudio staff

EXPOSE 8787

CMD /usr/lib/rstudio-server/bin/rserver --server-daemonize 0
`

**Comment by kevinushey** [comment]
Thanks -- I'm able to reproduce now. Here's an MRE that reproduces outside of RStudio:

```
library(Rfast)
dplyr <- asNamespace("dplyr")
print(dplyr)
```

Looks like `Rfast` installs some kind of print handler that gets stuck in an infinite loop when printing some package environments.

**Comment by lbergmans** [comment]
thanks @mpjdem ; this must be the issue in my case as well: the project where I had this issue was the first one where I used Rfast... (but I did not think about that..).
When I get back to the project, I'll get rid of Rfast (Real Fast)...

**Comment by jmcphers** [comment]
Leaving open for tracking. @kevinushey is there any change needed in RStudio, or are we at the mercy of `Rfast`?

**Comment by kevinushey** [comment]
We worked around at least one iteration of the problem here:

https://github.com/rstudio/rstudio/pull/6861

but ultimately, user code could trigger the `Rfast` method causing the R session to hang so there's not much we can do in that case.
